man/lldpctl.8
.\" Copyright (c) 2006 Pierre-Yves Ritschard <pyr@openbsd.org>
.\" Copyright (c) 2008 Vincent Bernat <bernat@luffy.cx>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: July 16 2008 $
.Dt LLDPCTL 8
.Os
.Sh NAME
.Nm lldpctl
.Nd control LLDP daemon
.Sh SYNOPSIS
.Nm
.Op Fl advwCN
.Op Fl L Ar location
.Op Fl P Ar policy
.Op Fl O Ar poe
.Op Fl o Ar poe
.Op Ar interface ...
.Sh DESCRIPTION
The
.Nm
program controls
.Xr lldpd 8
daemon.
.Pp
When no specific option is given,
.Nm
displays the list of discovered neighbors along with
some of their advertised capabilities. If some interfaces are given,
only those interfaces will be displayed.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
Enable more debugging information.
.It Fl v
Show
.Nm
version.
.It Fl a
Display all remote ports, including those hidden by the smart filter.
.It Fl w
Monitor neighbor changes. Each changed neighbor will be displayed.
.It Fl C
Display global configuration of
.Em lldpd
daemon.
.It Fl N
Make
.Em lldpd
update its information and send new LLDP PDU on all interfaces.
.It Fl f Ar format
Choose the output format. Currently
.Em plain ,
.Em xml
and
.Em keyvalue
formats are available. The default is
.Em plain.
.It Fl L Ar location
Enable the transmission of LLDP-MED location TLV for the given
interfaces. This option can be repeated several times to enable the
transmission of the location in several formats. Several formats are
accepted:
.Bl -tag -width "XX"
.It Em Coordinate based location
The format of
.Ar location
is
.Ar 1:48.85667N:2.2014E:117.47:m:1
The first digit is always
.Ar 1 .
It is followed by the latitude, a letter for the direction (
.Ar E
or
.Ar W
for East or West), the longitude and a letter for the direction (
.Ar N
or
.Ar S
). The next figure is the altitude. It can be expressed in meters (the
next letter is then
.Ar m
) or in floors (the letter should be
.Ar f
). The last digit is the datum. It can either be
.Ar 1
(WGS84),
.Ar 2
(NAD83) or
.Ar 3
(NAD83/MLLW).
.It Em Civic address
The location can be expressed as an address. The format of the
location is then
.Ar 2:FR:6:Commercial Rd:3:Roseville:19:4
The first digit is always
.Ar 2 .
The next two letters are the country code. Then, arguments are paired
to form the address. The first member of the pair is a digit
indicating the type of the second member. Here is the list of
valid types:
.Bl -tag -width "XXXX." -compact
.It Sy 0
Language
.It Sy 1
National subdivisions
.It Sy 2
County, parish, district
.It Sy 3
City, township
.It Sy 4
City division, borough, ward
.It Sy 5
Neighborhood, block
.It Sy 6
Street
.It Sy 16
Leading street direction
.It Sy 17
Trailing street suffix
.It Sy 18
Street suffix
.It Sy 19
House number
.It Sy 20
House number suffix
.It Sy 21
Landmark or vanity address
.It Sy 22
Additional location info
.It Sy 23
Name
.It Sy 24
Postal/ZIP code
.It Sy 25
Building
.It Sy 26
Unit
.It Sy 27
Floor
.It Sy 28
Room number
.It Sy 29
Place type
.It Sy 128
Script
.El
.It ECS ELIN
This is a numerical string using for setting up emergency call. The
format of the location is then the following:
.Ar 3:0000000911
where the first digit should be
.Ar 3
and the second argument is the ELIN number.
.El
.Pp
When setting a location for a given port, all previous locations are
erased. To erase all location, just use the empty string. There is
currently no way to get the location from the command line.
.Pp
.It Fl P Ar policy
Enable the transmission of LLDP-MED Network Policy TLVs for the given
interfaces. This option can be repeated several times to specify
different policies. Format (without spaces!):
.Pp
.Em App-Type
:
.Ar U
:
.Ar T
:
.Ar VLAN-ID
:
.Ar L2-Prio
:
.Ar DSCP
.Bl -tag -width "XX"
.It Ar App-Type
Valid application types (see ANSI/TIA-1057 table 12):
.Bl -tag -width "X." -compact
.It Sy 1
Voice
.It Sy 2
Voice Signaling
.It Sy 3
Guest Voice
.It Sy 4
Guest Voice Signaling
.It Sy 5
Softphone Voice
.It Sy 6
Video Conferencing
.It Sy 7
Streaming Video
.It Sy 8
Video Signaling
.El
.It Ar U
Unknown Policy Flag.
.Bl -tag -width "X." -compact
.It Sy 0
Network policy for the specified application type is defined.
.It Sy 1
Network policy for the specified application type is required by
the device but is currently unknown. This is used by Endpoint
Devices, not by Network Connectivity Devices.
.El
.It Ar T
Tagged Flag.
.Bl -tag -width "X." -compact
.It Sy 0
Untagged VLAN. In this case the VLAN ID and the Layer 2 Priority
are ignored and only the DSCP value has relevance.
.It Sy 1
Tagged VLAN.
.El
.It Ar VLAN-ID
IEEE 802.1q VLAN ID (VID). A value of 1 through 4094 defines a
VLAN ID. A value of 0 means that only the priority level is
significant.
.It Ar L2-Prio
IEEE 802.1d / IEEE 802.1p Layer 2 Priority, also known as Class of Service
(CoS), to be used for the specified application type.
.Bl -tag -width "X." -compact
.It Sy 1
Background
.It Sy 2
Spare
.It Sy 0
Best Effort (default)
.It Sy 3
Excellent Effort
.It Sy 4
Controlled Load
.It Sy 5
Video
.It Sy 6
Voice
.It Sy 7
Network Control
.El
.It Ar DSCP
DiffServ/Differentiated Services Code Point (DSCP) value as defined
in IETF RFC 2474 for the specified application type. Value: 0 (default
per RFC 2475) through 63. Note: The class selector DSCP values are
backwards compatible for devices that only support the old IP
precedence Type of Service (ToS) format. (See the RFCs for what
these values mean.)
.It Examples:
.Bl -tag -width "X." -compact
.It Sy 1:0:1:500:6:46
Voice (1): not unknown (0), tagged (1), VLAN-ID 500, l2 prio Voice (6), DSCP 46 (EF, Expedited Forwarding)
.It Sy 2:0:1:500:3:24
Voice Signaling (2): not unknown (0), tagged (1), VLAN-ID 500, l2 prio
Excellent Effort (3), DSCP 24 (CS3, Class Selector 3)
.El
.El
.Pp
.It Fl O Ar poe
Enable the transmission of LLDP-MED POE-MDI TLV for the given
interfaces. One can act as a PD (power consumer) or a PSE (power
provider). No check is done on the validity of the parameters while
LLDP-MED requires some restrictions:
.Bl -bullet
.It
PD shall never request more power than physical 802.3af class.
.It
PD shall never draw more than the maximum power advertised by PSE.
.It
PSE shall not reduce power allocated to PD when this power is in use.
.It
PSE may request reduced power using conservation mode
.It
Being PSE or PD is a global paremeter, not a per-port parameter.
.Nm
does not enforce this: a port can be set as PD or PSE. LLDP-MED also
requires for a PSE to only have one power source (primary or
backup). Again,
.Nm
does not enforce this. Each port can have its own power source. The
same applies for PD and power priority. LLDP-MED MIB does not allow
this kind of representation.
.El
.Pp
This option is distinct of
.Fl o
option. You may want to use both options at the same time.
.Pp
The format of this option is (without spaces):
.Pp
.Em type
:
.Ar source
:
.Ar priority
:
.Ar value
.Bl -tag -width "XX"
.It Ar type
Valid types are:
.Bl -tag -width "XXX." -compact
.It Sy PSE
Power Sourcing Entity (power provider)
.It Sy PD
Power Device (power consumer)
.El
.It Ar source
Valid sources are:
.Bl -tag -width "X." -compact
.It Sy unknown
Unknown
.It Sy primary
For PSE, the power source is the primary power source.
.It Sy backup
For PSE, the power source is the backup power source or a power
conservation mode is asked (the PSE may be running on UPS for
example).
.It Sy pse
For PD, the power source is the PSE.
.It Sy local
For PD, the power source is a local source.
.It Sy both
For PD, the power source is both the PSE and a local source.
.El
.It Ar priority
Four priorities are available:
.Bl -tag -width "X." -compact
.It Sy unknown
Unknown priority
.It Sy critical
Critical
.It Sy high
High
.It Sy low
Low
.El
.It Ar value
For PD, the power value is the total power in milliwatts required
by a PD device from the PSE device.
.El
.It Fl o Ar poe
Enable the transmission of Dot3 POE-MDI TLV for the given
interfaces. One can act as a PD (power consumer) or a PSE (power
provider). This option is distinct of the
.Fl O
option. You might want to use both. Contrary to LLDP-MED POE-MDI TLV,
Dot3 POE-MDI TLV are strictly per-port values.
.Pp
The format of this option is (without spaces):
.Pp
.Em type
:
.Ar supported
:
.Ar enabled
:
.Ar paircontrol
:
.Ar powerpairs
:
.Ar class
[ :
.Ar powertype
:
.Ar source
:
.Ar priority
:
.Ar requested
:
.Ar allocated
]
.Bl -tag -width "XX"
.It Ar type
Valid types are:
.Bl -tag -width "XXX." -compact
.It Sy PSE
Power Sourcing Entity (power provider)
.It Sy PD
Power Device (power consumer)
.El
.It Ar powerpairs
Valid values are:
.Bl -tag -width "X." -compact
.It Sy signal
The signal pairs only are in use.
.It Sy spare
The spare pairs only are in use.
.El
.It Ar class
Five classes are available:
.Bl -tag -width "X." -compact
.It Sy 1
class 0
.It Sy 2
class 1
.It Sy 3
class 2
.It Sy 4
class 3
.It Sy 5
class 4
.It Sy 0
no class
.El
.El
.Pp
.Ar supported ,
.Ar enabled
and
.Ar paircontrol
can be set to to 0 or 1.
.Ar supported
means that MDI power is supported on the given port.
.Ar enabled
means that MDI power is enabled on the given port.
.Ar paircontrol
is used to indicate if the pair selection can be controlled on the
given port.
.Pp
.Ar powertype ,
.Ar source ,
.Ar priority
(and remaining values) are optional. They are only requested in
conformance with 802.3at.
.Ar type
should be either 1 or 2. For source, use one of the following values:
Valid sources are:
.Bl -tag -width "X." -compact
.It Sy 0
Unknown
.It Sy 1
For PD, the power source is the PSE. For PSE, the power source is the
primary power source.
.It Sy 2
For PD, the power source is a local source. For PSE, the power source
is the backup power source or a power conservation mode is asked (the
PSE may be running on UPS for example).
.It Sy 3
For PD, the power source is both the PSE and a local source. For PSE,
this value should not be used.
.El
For
.Ar priority ,
see what is done for LLDP-MED MDI/POE.
.Ar requested
and
.Ar allocated
are respectively the PD requested power value and the PSE allocated
power value. This should be expressed in milliwatts.
.El
.Sh FILES
.Bl -tag -width "/var/run/lldpd.socketXX" -compact
.It /var/run/lldpd.socket
Unix-domain socket used for communication with
.Xr lldpd 8 .
.El
.Sh SEE ALSO
.Xr lldpd 8
.Sh AUTHORS
.An -nosplit
The
.Nm
program was written by
.An Vincent Bernat Aq bernat@luffy.cx .


NEWS
lldpd (0.6.1)
  * Features:
    + Provide liblldpctl.so, a library to interface with lldpd. The
      documentation is provided through Doxygen. See src/lib/lldpctl.h
      which contains all the exported functions.
    + Make lldpctl uses liblldpctl.so.
    + Add a "watch" option to lldpctl to monitor neighbor changes.
    + Add the possibility to display the current configuration of
      lldpd with lldpctl. Also add the possibility to reset the
      current transmit delay.

lldpd (0.6)
  * Features:
    + Allow lldpctl to display hidden ports.
    + Add a switch to specify interfaces to use to get chassis ID.
    + Support for multiple management addresses and IPv6 management
      addresses. Contributed by João Valverde.
  * Global changes:
    + Switch to libevent. See README.md for details.
    + Partial rewrite of the SNMP part. Less code.
    + Unit tests for SNMP.
    + Major rewrite of the protocol between lldpd and lldpctl. Less
      code.
  * Fixes:
    + Several small SNMP fixes (discovered by unit tests).

lldpd (0.5.7)
 * Fixes:
    + Configure issue with NetSNMP and some linkers
    + Fix infinite loop for the receive part: on certain conditions,
      lldpd will stop sending packets and stop updating local data.

lldpd (0.5.6)
  * Features:
    + Send and receive native VLAN TLV with CDP
    + Add a whitelist for some drivers (currently: dsa and veth)
  * Fixes:
    + Compilation issues with NetSNMP 5.7 (and with earlier versions too)
    + Small optimization of BPF filter

lldpd (0.5.5)
  * Features:
    + Support for PPVID and PI Dot1 TLV, thanks to Shuah Khan.
    + Extend whitelist with possibility to blacklist.
  * Fixes:
    + Key/value output was incorrect when a dot was present in
      interface names. This is fixed but it is preferable to use XML
      output since the parsing is more difficult in this case.
    + Only grab DMI information once. Only uses DMI for x86 platform.
    + Padding issues with socket protocol. This introduces a change in
      the socket protocol!
    + Fix a segfault when neither /etc/os-release nor lsb_release
      are available.

lldpd (0.5.4)
  * Features:
    + Get OS information from /etc/os-release if available. Patch from
      Michael Tremer.
    + Add a flag to specify which interfaces lldpd should listen to.

lldpd (0.5.3)
  * Fixes:
    + Allow root to change configuration of lldpd when lldpctl has suid set.
  * Features:
    + Handle Dot3 POE-MDI TLV (802.3af and 802.3at).
    + Allow to set Dot3 POE-MDI from lldpctl.

lldpd (0.5.2)
  * Features:
    + More flexible smart mode and new default. Manual page has been updated.
    + Add a "receive-only" mode with "-r" switch.

lldpd (0.5.1)
  * Features:
    + Allow to force a protocol even when no peer for this protocol is
      detected.
    + Add a smart mode that allows to discard bogus port information,
      for example CDP packets that are flooded through a switch that
      does not support CDP.
    + Allow to set LLDP-MED network policy from lldpctl, thanks to a patch from
      Philipp Kempgen.
    + Allow to set LLDP-MED POE-MDI from lldpctl.
    + Add a summary of available options in "lldpd -h" and "lldpctl -h",
      thanks to a patch from Jorge Boncompte.
    + Add a new output (keyvalue) for lldpctl.
    + Listen on VLAN using an appropriate BPF filter, VLAN
      decapsulation. Older "listen on vlan" feature is discarded. See
      README for more information on the new feature.
    + Use output of lsb_release if available for system description.

  * Fixes:
    + Ignore interface with no queue. It should filter out interfaces
      like "vnet0" that would fail if we try to send something on them.
    + Don't check CDP checksums (not really a fix but it appears that
      Cisco checksum have some difficult corner cases).

lldpd (0.5.0)

  * Features:
    + lldpd can now handle several systems on the same port. This
      modification also allows to speak to a switch using CDP and LLDP
      for example.
    + The way that lldpd gathers information for each port has been
      abstracted. This should allow to support more systems (BSD for
      examples) or switch cores in the future. Sending/receive support
      is also abstracted.
    + Add "-k" switch to avoid to emit too much information on running
      kernel.
    + Support of ifAlias with kernel >= 2.6.28
    + Lot of portability stuff. lldpd can now be compiled on RHEL
      2.1. Still Linux-only though.
    + Add an option to specify AgentX socket (-X).
    + Add some unit tests
    + lldpctl has been reworked; it is now able to output data in XML
      format for easier parsing. Patches were provided by Andreas
      Hofmeister.

  * Bug fixes:
    + Fix EDP VLAN handling
    + Silent warnings about bridge stuff.
    + Copy /etc/localtime into chroot before starting lldpd daemon to
      ensure correct timestamps for logs.

lldpd (0.4.1)

  * Fix EDP handling when there is no VLAN
  * Fix CDP version to not always be 1
  * Misc fix:
     + incorrect number of arguments for a LLOG_INFO call
     + fix SNMP last change in case this change occurs before start time

lldpd (0.4)

  * Rewrite of packet builder and parser to be able to cope with
    architecture that cannot do unaligned read. For decoder, we don't
    cast structures any more since they can be unaligned. For encoder,
    we use memcpy through the use of macro that build packets step by
    step.

lldpd (0.3.2)

  * Fix LLDP-MED support

lldpd (0.3.1)

  * Misc fixes, including memory leaks

lldpd (0.3)

  * Initial support of LLDP-MED
  * Fix for bridge detection (don't send bridge ioctl on random interfaces)
  * For bonded devices, get the real hardware address. For inactive
    slaves, transmit using a random MAC address.

lldpd (0.2.1)

  * Fix a syntax error in manual page
  * Fix open() calls

lldpd (0.2)

  * Add privilege separation
  * Add FDP support
  * Support CDP encapsulated into native VLAN
  * Various fixes

lldpd (0.1)


src/client/lldpctl.c
/* -*- mode: c; c-file-style: "openbsd" -*- */
/*
 * Copyright (c) 2008 Vincent Bernat <bernat@luffy.cx>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <arpa/inet.h>

#include "../log.h"
#include "../ctl.h"
#include "client.h"

static void		 usage(void);

#ifdef HAVE___PROGNAME
extern const char	*__progname;
#else
# define __progname "lldpctl"
#endif


static void
usage(void)
{
	fprintf(stderr, "Usage:   %s [OPTIONS ...] [INTERFACES ...]\n", __progname);
	fprintf(stderr, "Version: %s\n", PACKAGE_STRING);

	fprintf(stderr, "\n");

	fprintf(stderr, "-d          Enable more debugging information.\n");
	fprintf(stderr, "-a          Display all remote ports, including hidden ones.\n");
	fprintf(stderr, "-w          Watch for changes.\n");
	fprintf(stderr, "-C          Display global configuration of lldpd.\n");
	fprintf(stderr, "-N          Make lldpd transmit LLDP PDU now.\n");
	fprintf(stderr, "-f format   Choose output format (plain, keyvalue or xml).\n");
	fprintf(stderr, "-L location Enable the transmission of LLDP-MED location TLV for the\n");
	fprintf(stderr, "            given interfaces. Can be repeated to enable the transmission\n");
	fprintf(stderr, "            of the location in several formats.\n");
	fprintf(stderr, "-P policy   Enable the transmission of LLDP-MED Network Policy TLVs\n");
	fprintf(stderr, "            for the given interfaces. Can be repeated to specify\n");
	fprintf(stderr, "            different policies.\n");
	fprintf(stderr, "-O poe      Enable the transmission of LLDP-MED POE-MDI TLV\n");
	fprintf(stderr, "            for the given interfaces.\n");
	fprintf(stderr, "-o poe      Enable the transmission of Dot3 POE-MDI TLV\n");
	fprintf(stderr, "            for the given interfaces.\n");

	fprintf(stderr, "\n");

	fprintf(stderr, "see manual page lldpctl(8) for more information\n");
	exit(1);
}

struct cbargs {
	int argc;
	char **argv;
	struct writer *w;
};

void
watchcb(lldpctl_conn_t *conn,
    lldpctl_change_t type,
    lldpctl_atom_t *interface,
    lldpctl_atom_t *neighbor,
    void *data)
{
	int ch, i;
	struct cbargs *args = data;
	optind = 0;
	while ((ch = getopt(args->argc, args->argv, LLDPCTL_ARGS)) != -1);
	if (optind < args->argc) {
		for (i = optind; i < args->argc; i++)
			if (strcmp(args->argv[i],
				lldpctl_atom_get_str(interface,
				    lldpctl_k_interface_name)) == 0)
				break;
		if (i == args->argc)
			return;
	}
	switch (type) {
	case lldpctl_c_deleted:
		tag_start(args->w, "lldp-deleted", "LLDP neighbor deleted");
		break;
	case lldpctl_c_updated:
		tag_start(args->w, "lldp-updated", "LLDP neighbor updated");
		break;
	case lldpctl_c_added:
		tag_start(args->w, "lldp-added", "LLDP neighbor added");
		break;
	default: return;
	}
	display_interface(conn, args->w, 1, interface, neighbor);
	tag_end(args->w);
}

int
main(int argc, char *argv[])
{
	int ch, debug = 1;
	char *fmt = "plain";
	int action = 0, hidden = 0, watch = 0, configuration = 0, now = 0;
	lldpctl_conn_t *conn;
	struct cbargs args;

	/* Get and parse command line options */
	while ((ch = getopt(argc, argv, LLDPCTL_ARGS)) != -1) {
		switch (ch) {
		case 'h':
			usage();
			break;
		case 'd':
			debug++;
			break;
		case 'v':
			fprintf(stdout, "%s\n", PACKAGE_VERSION);
			exit(0);
			break;
		case 'a':
			hidden = 1;
			break;
		case 'f':
			fmt = optarg;
			break;
		case 'L':
		case 'P':
		case 'O':
		case 'o':
			action = 1;
			break;
		case 'w':
			watch = 1;
			break;
		case 'C':
			configuration = 1;
			break;
		case 'N':
			now = 1;
			break;
		default:
			usage();
		}
	}

	log_init(debug, __progname);

	if ((action != 0) &&
	    (getuid() != geteuid() || getgid() != getegid())) {
		fatalx("mere mortals may not do that, admin privileges are required.");
	}

	conn = lldpctl_new(NULL, NULL, NULL);
	if (conn == NULL) exit(EXIT_FAILURE);

	args.argc = argc;
	args.argv = argv;
	if (watch) {
		if (lldpctl_watch_callback(conn, watchcb, &args) < 0) {
			LLOG_WARNX("unable to watch for neighbors. %s",
			    lldpctl_last_strerror(conn));
			exit(EXIT_FAILURE);
		}
	}

	do {
		if (strcmp(fmt, "plain") == 0) {
			args.w = txt_init(stdout);
		} else if (strcmp(fmt, "keyvalue") == 0) {
			args.w = kv_init(stdout);
		}
#ifdef USE_XML
		else if (strcmp(fmt,"xml") == 0 ) {
			args.w = xml_init(stdout);
		}
#endif
		else {
			args.w = txt_init(stdout);
		}

		if (action) {
			modify_interfaces(conn, argc, argv, optind);
		} else if (watch) {
			if (lldpctl_watch(conn) < 0) {
				LLOG_WARNX("unable to watch for neighbors. %s",
				    lldpctl_last_strerror(conn));
				watch = 0;
			}
		} else if (configuration) {
			display_configuration(conn, args.w);
		} else if (now) {
			lldpctl_atom_t *config = lldpctl_get_configuration(conn);
			if (config == NULL) {
				LLOG_WARNX("unable to get configuration from lldpd. %s",
					lldpctl_last_strerror(conn));
			} else {
				if (lldpctl_atom_set_int(config,
					lldpctl_k_config_tx_interval, -1) == NULL) {
					LLOG_WARNX("unable to ask lldpd for immediate retransmission. %s",
						lldpctl_last_strerror(conn));
				} else
					LLOG_INFO("immediate retransmission requested successfuly");
				lldpctl_atom_dec_ref(config);
			}
		} else {
			display_interfaces(conn, args.w,
			    hidden, argc, argv);
		}
		args.w->finish(args.w);
	} while (watch);

	lldpctl_release(conn);
	return EXIT_SUCCESS;
}
